### DISCUSSION TITLE
Merged file is bigger, with smaller cover art

### DISCUSSION THREAD

**Discussion by AudiobookListener** [yermak/AudioBookConverter#311] [General]
Was merging 39 M4B chapter files (1.01GB total) with AudioBookConverter, and noticed that the resulting file was larger (1.02GB). Not much, but strange as the 39 chapter files all had 66K of cover art. The the merged file seems to only have one copy but strangly the cover art size is now 50K.  Ran the same merge with inAudible and the resulting file size was a bit smaller than the individual chapter (1.00GB) files as expected.

Precise data for analysis
Cover art 39*66K=2574K/.33MB, 39*50K=1950K/.25MB
(1,084,541,330 bytes) 1084.5MB (on Disk 1,084,542,976 bytes) Merged by inAudible, 30 second merge time
(1,087,196,048 bytes) 1087.2MB (on Disk 1,087,275,008 bytes) 39 M4B files, each with 66K cover art
(1,096,307,629 bytes) 1096.1MB (on Disk 1,096,310,784 bytes)  Merged by AudiobookBookConverter, cover size now 50K, 80 second merge time

Seemingly, this paid product took 3x longer and expanded the result, whereas the freeware was faster and produced a smaller file.

http://malich.ru/duplicate_searcher which has the option to compare media files minus the metadata found merged audio not equal.

Perhaps we can make some sense of this? Thanks, -AudiobookListener

**Comment by yermak** [comment]
@AudiobookListener,
Thank you very much for feedback, i'm not sure if this a bug or general discussion.

First important thing to know when you merge existing M4B files - there is no real encoding happens in most of the cases, it's just copying files on the disk and muxing them. In most real life re-encoding scenarios, this time is very minor comparing to real time of actual re-encoding.
Secondly 3 times of time difference in this corner case scenario is easy to explain and works as designed: as all manipulation with audio files are happening in "TEMP" directory, so files are copied there first, this is generally useful when files are stored on network drive like NAS. And another round of copping happens due to use of parameter "-moveflag faststart"
Here is extract from encoder documentation:
```
By default the MP4 muxer writes the 'moov' atom after the audio stream ('mdat' atom) at the end of the file. This results in the user requiring to download the file completely before playback can occur. Relocating this moov atom to the beginning of the file can facilitate playback before the file is completely downloaded by the client.
-movflags faststart
Run a second pass moving the index (moov atom) to the beginning of the file. This operation can take a while, and will not work in various situations such as fragmented output, thus it is not enabled by default.
```
I do believe spending few extra-second during re-encoding is better then waiting while file loads during playback.

This could explain size difference as well, but i'm not sure about that w/o testing.

If you interested in further discussion share your files by link in cloud and send to my nick name @ gmail.com.
I will run a test have a look.

Please feel free to share link to inAudible (as i'm not sure what exact software you refer to), as AudiobookConverter is opensource software i'm really interested in community help to make it better.

P.S. On a separate note: to compare performance speed please select testing set where real re-encoding will happen (i.e. mp3 to m4b or m4b to mp3), i'm sure you will not be disappointed by AudiobookConverter performance.

**Comment by yermak** [comment]
Converted issue to discussion

**Reply by AudiobookListener** [reply]
So, more testing.... Not sure why the cover art shrank from 66K to 50K on the merge, but found a solution for the file-size expansion. It is resolved with the MP3tag "Optimize MP4" utility. It shrinks the combined MP4 down to a competitive size. Here's a discussion of the utility...

**Question: Does the MP3tag optimize mp4 utility  have to do with moving the metadata block from the end of the file to the beginning to make the files "fast start streaming" compatible? 

Answer: "This function removes all "free" MP4-atoms and relocates the "moov" and "tags" MP4-Atoms in front of the audio data. This reduces the size of the files and may solve some problems with not displayed tags in other programs."**

It would appear that AudioBookConverter does not optimize the MP4 metadata as it is merging the files. In the case of 39 files totaling 1GB, it added nearly 12MB of bloat. I suspect the amount of bloat is proportional to the amount of data and the number of chapters merged. -AudiobookListener

**Comment by yermak** [comment]
Thanks for reporting this.
I was able to reproduce it, indeed there is a strange behaviour (bug) caused by mp4art from mp4v2 tool i used.
After audio is converted, merged and optimised application adds art work as separate action, and after this file become fragmented again, in my case it was 16MB of overhead added.
After I re-optimised file again new video size became orginal record + jpg size + 21 byte. 

This was completely unexpected, i'm not sure if this a bug introduced recently, but as workaround is known i will fix it next version.

**Comment by yermak** [comment]
As for cover art size - sometimes, I have to re-encode art file, especially if there are many instances of the file embaded into audiofile.
I don't think there is a visual difference, so this i minor issue.

**Comment by AudiobookListener** [comment]
The cover art always seems to shrink to 50K, this is a minor issue...for me.

**Reply by AudiobookListener** [reply]
Hello again, using the new version (from this week), and just merged 2 parts of a single M4B audiobook. Strangly, the resulting M4B contained 6 pieces of cover art. The Art the 2 files shared, and 5 others, some of which were from other books I've cleaned up. Perhaps the temp directory is shared with another program or isn't getting cleaned up?

**Comment by yermak** [comment]
Did you check what what is ArtWork tab before you pressed Start?
If you they were not there it is very strange and needs to be investigated, if you so them there - there is explanation for such behaviour.

Program import images from files added and any images from all directories recursively where source files were limiting with 10 images.
This is needed for very typical case where mp3s are in the directory and coverarts are stored separately in the same directory.

However (checked the source code), there is a chance of clash filenames between images created - which i need to fix anyway.
If you confirm have you seen them in artwork tab that would be helpful, also in the first case it will be very difficult to reproduce this issue.
